### PR TITLE
WPC-103 Subscription filters

### PR DESCRIPTION
**Pull request by antkiewiczk**
Working filters:
- by member plan
- by payment periodicity
- by payment method
- by auto renew
- by deactivation reason
- by deactivation date (range) 
- by paid until (range)
- by start at (range)

Not implemented:
- by user address

Also: generate and download csv file

### CODE REVIEW

File: examples/api/src/index.ts [hunk 1]
```diff
@@ -221,7 +221,7 @@ async function asyncMain() {
   }
 
   const paymentProviders = []
-
+  console.log('process.env.STRIPE_SECRET_KEY', process.env.STRIPE_SECRET_KEY)
```
[the remark this answers is not in the document]
Comment (tomaszdurka) [reply examples/api/src/index.ts:224]: :)

File: packages/editor/src/client/routes/subscriptionList.tsx [hunk 2]
```diff
@@ -145,14 +202,172 @@ export function SubscriptionList() {
...
+                label: 'No'
+              }
+            ]}
+            value={filter.deactivationReason}
+            block
+            placement="auto"
+            onChange={value => updateFilter({autoRenew: value || undefined})}
+          />
+        </FormGroup>
+        {/*  hide for now
```
[the remark this answers is not in the document]
Comment (antkiewiczk) [reply packages/editor/src/client/routes/subscriptionList.tsx]: done

File: packages/api/src/graphql/mutation.private.ts [hunk 3]
```diff
@@ -479,7 +479,7 @@ export const GraphQLAdminMutation = new GraphQLObjectType<undefined, Context>({
 
         const subscription = await dbAdapter.subscription.createSubscription({input})
         if (!subscription) throw new Error('Subscription not created.')
-
+        console.log('subscription', subscription)
```
[the remark this answers is not in the document]
Comment (antkiewiczk) [reply packages/api/src/graphql/mutation.private.ts]: done

File: examples/api/.env.example [hunk 4]
```diff
@@ -21,9 +21,9 @@ OAUTH_WEPUBLISH_REDIRECT_URL=http://localhost:4000/auth/wepublish
 
 JWT_SECRET_KEY=ThisIsSuperSecret
 
-#STRIPE_SECRET_KEY=
-#STRIPE_CHECKOUT_WEBHOOK_SECRET=
-#STRIPE_WEBHOOK_SECRET=
+STRIPE_SECRET_KEY=some_test_value
```
[the remark this answers is not in the document]
Comment (antkiewiczk) [reply examples/api/.env.example]: done

File: packages/api-db-mongodb/src/db/subscription.ts [hunk 5]
```diff
@@ -215,20 +214,38 @@ export class MongoDBSubscriptionAdapter implements DBSubscriptionAdapter {
       textFilter.$and = []
     }
 
-    if (filter?.startsAt !== undefined) {
-      const {comparison, date} = filter.startsAt
+    if (filter?.startsAtFrom !== undefined) {
```
[the remark this answers is not in the document]
Comment (antkiewiczk) [reply packages/api-db-mongodb/src/db/subscription.ts]: done

File: packages/api/src/graphql/mutation.private.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (antkiewiczk) [reply packages/api/src/graphql/mutation.private.ts]: done

File: examples/api/src/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (antkiewiczk) [reply examples/api/src/index.ts]: done

File: packages/api-db-mongodb/src/db/subscription.ts [hunk 6]
```diff
@@ -243,25 +260,68 @@ export class MongoDBSubscriptionAdapter implements DBSubscriptionAdapter {
...
+    // }
+
     const [totalCount, subscriptions] = await Promise.all([
       this.subscriptions.countDocuments(textFilter, {
         collation: {locale: this.locale, strength: 2}
       } as MongoCountPreferences), // MongoCountPreferences doesn't include collation
 
       this.subscriptions
         .aggregate([], {collation: {locale: this.locale, strength: 2}})
+        // how to get related user address?
```
Comment (penk85) [inline packages/api-db-mongodb/src/db/subscription.ts:318]: Does this need to be here?

File: packages/api/src/db/subscription.ts [hunk 7]
```diff
@@ -52,10 +52,20 @@ export enum SubscriptionSort {
 
 export interface SubscriptionFilter {
   readonly startsAt?: DateFilter
```
Comment (penk85) [inline packages/api/src/db/subscription.ts:54]: Is the `startsAt` unused now? Maybe can be removed.

File: packages/api/src/db/subscription.ts [hunk 8]
```diff
@@ -52,10 +52,20 @@ export enum SubscriptionSort {
...
+  readonly startsAtFrom?: DateFilter
+  readonly startsAtTo?: DateFilter
   readonly paidUntil?: DateFilter
```
Comment (penk85) [inline packages/api/src/db/subscription.ts:57]: Same question here.

File: packages/editor/src/client/panel/ExportSubscriptionsCsvModal.tsx [hunk 9]
```diff
@@ -24,24 +43,43 @@ export function SubscriptionAsCsvModal() {
...
+              contentType: 'text/csv;charset=utf-8;'
+            })
+          }>
+          {t('userList.panels.download')}
+        </IconButton>
+        <IconButton
+          appearance="primary"
+          icon={<Icon size="lg" icon="copy" />}
+          disabled={!subscriptionsCsvData?.subscriptionsAsCsv}
+          onClick={() => navigator.clipboard.writeText(subscriptionsCsvData!.subscriptionsAsCsv!)}>
```
Comment (penk85) [inline packages/editor/src/client/panel/ExportSubscriptionsCsvModal.tsx:68]: Is the `!` necessary here? Can you get around it with `|| ' '` instead?

Comment (antkiewiczk) [reply packages/editor/src/client/panel/ExportSubscriptionsCsvModal.tsx:68]: yeah these kind of assertions are all through the project and my thought was that we should get rid of them all together

File: packages/api/src/db/subscription.ts (hunk 8 above)
Comment (antkiewiczk) [reply packages/api/src/db/subscription.ts:57]: some of them are used in other methods, not related to the filtering itself so I decided to just leave them as they are (not wanting to break anything)

File: packages/api-db-mongodb/src/db/subscription.ts (hunk 6 above)
Comment (antkiewiczk) [reply packages/api-db-mongodb/src/db/subscription.ts]: removed

File: packages/api/src/db/subscription.ts (hunk 8 above)
Comment (penk85) [reply packages/api/src/db/subscription.ts:57]: Sounds reasonable

**Review by penk85 — APPROVED** [review]
Looks great :+1:

File: packages/api/src/graphql/query.private.ts [hunk 10]
```diff
@@ -336,21 +336,21 @@ export const GraphQLQuery = new GraphQLObjectType<undefined, Context>({
...
-
-        return await dbAdapter.subscription.getSubscriptions({
+        const toReturn = await dbAdapter.subscription.getSubscriptions({
           filter,
           sort,
           order,
           cursor: InputCursor(after, before),
           limit: Limit(first, last, skip)
         })
+        return toReturn
```
Comment (tomaszdurka) [inline packages/api/src/graphql/query.private.ts:346]: That can be reverted.

File: packages/api/src/graphql/query.private.ts [hunk 11]
```diff
@@ -360,16 +360,16 @@ export const GraphQLQuery = new GraphQLObjectType<undefined, Context>({
...
+          const listResult: ConnectionResult<Subscription> = await dbAdapter.subscription.getSubscriptions(
+            {
+              filter,
               limit: Limit(100),
               sort: SubscriptionSort.ModifiedAt,
+              cursor: InputCursor(afterCursor ?? undefined),
               order: SortOrder.Descending
-            })
+            }
+          )
```
Comment (tomaszdurka) [inline packages/api/src/graphql/query.private.ts:372]: It's a bit strange formatting. Do you use our eslint?

File: packages/editor/src/client/api/subscription.graphql [hunk 12]
```diff
@@ -1,77 +1,92 @@
 fragment FullSubscription on Subscription {
-    id
-    createdAt
-    modifiedAt
+  id
```
Comment (tomaszdurka) [inline packages/editor/src/client/api/subscription.graphql:2]: What is the story with formatting? Are you using other formatting then we had? We should all use the same agreed so we don't go back and forth with it.

**Review by tomaszdurka — CHANGES_REQUESTED** [review]
small things

**Review by tomaszdurka — APPROVED** [review]

**Review by Itrulia — DISMISSED** [review]
Amazing!

**Review by elias-summermatter — DISMISSED** [review]

**Review by penk85 — DISMISSED** [review]

**Review by Itrulia — APPROVED** [review]

**Review by tomaszdurka — APPROVED** [review]

**Pull request merged**